src/sqlfluff/dialects/dialect_postgres.py

```diff
@@ -402,7 +402,34 @@ class DatatypeSegment(BaseSegment):
         ),
         Sequence(
             OneOf(
+                # numeric types
+                "SMALLINT",
+                "INTEGER",
+                "INT2",
+                "INT4",
+                "INT8",
+                "BIGINT",
+                "REAL",
                 Sequence("DOUBLE", "PRECISION"),
+                "SMALLSERIAL",
+                "SERIAL",
+                "SERIAL2",
+                "SERIAL4",
+                "SERIAL8",
+                "BIGSERIAL",
+                # numeric types [(precision)]
+                Sequence(
+                    OneOf("FLOAT", "FLOAT4", "FLOAT8"),
+                    Bracketed(Ref("NumericLiteralSegment"), optional=True),
+                ),
+                # numeric types [precision ["," scale])]
+                Sequence(
+                    OneOf("DECIMAL", "NUMERIC"),
+                    Bracketed(
+                        Delimited(Ref("NumericLiteralSegment")),
+                    ),
+                    optional=True,
```

Comment:
This optional is in the wrong place. if you refer to the parse tree you can see the current grammar doesn't handle a lone DECIMAL/NUMERIC
```suggestion
                    Bracketed(
                        Delimited(Ref("NumericLiteralSegment")),
                    optional=True,
                    ),
```

Reply:
In general this is the right approach though and the parse tree looks fine 👍 

Couple notes:
- DataTypeIdentifier: FYI This is a general regex matcher for things potential datatypes. The purpose of this is to catch user defined data types. It should exclude Reserved Keywords, but doesn't atm due to the type grammar not being fully implemented as we discussed in the original issue.
- Brackets after data types: if you refer to line 456 you can see that there is a bracketed segment, which I think was originally put in as a general way to get brackets after a datatype. But means that weird types can parse:
![image](https://user-images.githubusercontent.com/80432516/146420155-fa2ee84d-60d1-423b-9437-ad38657dd76f.png)
Basically we'll need to be more specific with the other datatypes as to which can and can't have brackets after them (like you did for FLOAT)

There aren't that many other types in https://www.postgresql.org/docs/14/datatype.html so it's easier to add them all in this PR and we can do the tidy up once rather than splitting this up.